toolchain/check/subst.cpp

```diff
@@ -0,0 +1,241 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "toolchain/check/subst.h"
+
+#include "toolchain/check/eval.h"
+#include "toolchain/sem_ir/copy_on_write_block.h"
+#include "toolchain/sem_ir/ids.h"
+
+namespace Carbon::Check {
+
+namespace {
+
+// Information about an instruction that we are substituting into.
+struct WorklistItem {
+  // The instruction that we are substituting into.
+  SemIR::InstId inst_id;
+  // Whether the operands of this instruction have been added to the worklist.
+  bool is_expanded : 1;
+  // The index of the worklist item to process after we finish updating this
+  // one. For the final child of an instruction, this is the parent. For any
+  // other child, this is the index of the next child of the parent. For the
+  // root, this is -1.
+  int next_index : 31;
+};
+
+// A list of instructions that we're currently in the process of substituting
+// into. For details of the algorithm used here, see `SubstConstant`.
+class Worklist {
+ public:
+  auto Get(int index) -> WorklistItem& { return worklist[index]; }
+  auto Size() -> int { return worklist.size(); }
+  auto Push(SemIR::InstId inst_id) -> void {
+    worklist.push_back({.inst_id = inst_id,
+                        .is_expanded = false,
+                        .next_index = static_cast<int>(worklist.size() + 1)});
+    CARBON_CHECK(worklist.back().next_index > 0) << "Constant too large.";
+  }
+  auto Pop() -> SemIR::InstId { return worklist.pop_back_val().inst_id; }
+
+ private:
+  // Constants can get pretty large, so use a large worklist. This should be
+  // about 4KiB, which should be small enough to comfortably fit on the stack,
+  // but large enough that it's unlikely that we'll need a heap allocation.
+  llvm::SmallVector<WorklistItem, 512> worklist;
+};
+
+}  // namespace
+
+// Pushes the specified operand onto the worklist.
+static auto PushOperand(Context& context, Worklist& worklist,
```

Comment:
Had you considered making these as members of a class rather than passing `Context`+`Worklist` as an argument?

Reply:
I thought about putting them on `Worklist`, but I like the separation between the worklist and these operations. I've not really thought about adding a class to hold the context and the worklist, and having now considered it I think it's probably not worthwhile -- passing them both around seems like less work than maintaining a separate declaration and definition of all the functions -- but I find this pretty borderline and would be happy to change it if you feel more strongly.